### PR TITLE
[Rich text editor] Add full screen mode

### LINKED ISSUE #1259: Editing multi-line message creates extra new lines
### Steps to reproduce

1. Send a multi-line message
2. Edit the message just sent
3. Send the message

#### What did you expect?
The edited message is the same as the original message

#### What happened instead?
The edited message contains extra newlines

https://github.com/vector-im/element-x-android/assets/4940864/5b525e32-d332-4ea8-957e-a07c0470b5dd

### Your phone model

_No response_

### Operating system version

Android 13

### Application version and app store

Element X 0.1.7-nightly (40001070)

### Homeserver

_No response_

### Will you send logs?

No

### Are you willing to provide a PR?

No

### LINKED ISSUE #1293: Formatting buttons get squashed after typing a long message
### Steps to reproduce

0. Ensure rich text editor is enabled
1. Open the text formatting menu
2. Start composing a message
3. Type many lines of text

#### What did you expect?

The formatting buttons are always visible

#### What happened instead?

The formatting buttons get squashed

https://github.com/vector-im/element-x-android/assets/4940864/d84d06d3-69f0-4f50-9df4-e4a52b857e30

### Your phone model

_No response_

### Operating system version

_No response_

### Application version and app store

Element X nightly 2022-06-13

### Homeserver

_No response_

### Will you send logs?

No

### Are you willing to provide a PR?

No

### LINKED ISSUE #1394: RTE: Bad things happen when you send emoji
### Steps to reproduce

* In RTE, send some formatting, then an emoji
* formatting gets lost
* linebreaks get lost too?
* possibly other badness shown on the video below

Video taken from etherhaus [here](https://matrix.to/#/!aCvOWKNDeXXpPgquKq:matrix.org/$Bq2dQ8rPbyJdtTeZ1UiCKKn__fDYhXlkqSy1DK81rQM?via=matrix.org&via=envs.net&via=tchncs.de)

https://github.com/vector-im/element-x-android/assets/1294269/739c628b-cf8a-477b-9e74-70ed81033dd4

#### What did you expect?

ability to mix emoji & formatted text in RTE

#### What happened instead?

"Text gains life"

### Your phone model

_No response_

### Operating system version

_No response_

### Application version and app store

0.2.0

### Homeserver

_No response_

### Will you send logs?

No

### Are you willing to provide a PR?

No

### PR DESCRIPTION
**Pull request by jonnyandrew**
## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

- Add full screen mode for the rich text editor (RTE). When text formatting options are enabled, the editor can be dragged to full screen.
- Remove `ConstraintLayout` from `textcomposer` module, now made much simpler now the RTE supports being called in multiple layouts matrix-org/matrix-rich-text-editor#822

To follow in a later PR:
- Refactor the large `TextComposer.kt` file and split UI components into multiple files. 

## Motivation and context

- Part of vector-im/element-meta#1973
- Includes design from #1315
- Fixes #1293 (through new layout)
- Fixes #1394 (through inclusion of matrix-org/matrix-rich-text-editor#824)
- Fixes #1259 (through inclusion of matrix-org/matrix-rich-text-editor#820)

## Screenshots / GIFs

[Screen_recording_20230928_144233.webm](https://github.com/vector-im/element-x-android/assets/4940864/86324cea-b4e8-4959-b86f-b6c624ff1873)

## Checklist

- [ ] Changes have been tested on an Android device or Android emulator with API 23
- [x] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/vector-im/element-x-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/vector-im/element-x-android/blob/develop/CONTRIBUTING.md#changelog
- [x] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR

### CODE REVIEW

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/messagecomposer/MessageComposerState.kt [hunk 1]
```diff
@@ -17,12 +17,13 @@
...
 
 import androidx.compose.runtime.Immutable
+import androidx.compose.runtime.Stable
 import io.element.android.features.messages.impl.attachments.Attachment
 import io.element.android.libraries.textcomposer.MessageComposerMode
 import io.element.android.wysiwyg.compose.RichTextEditorState
 import kotlinx.collections.immutable.ImmutableList
 
-@Immutable
+@Stable
```
Comment (jonnyandrew) [inline features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/messagecomposer/MessageComposerState.kt:26]: `richTextEditorState` is mutable

File: libraries/textcomposer/impl/src/main/kotlin/io/element/android/libraries/textcomposer/TextComposer.kt [hunk 2]
```diff
@@ -105,319 +105,404 @@ fun TextComposer(
...
 
-        // Placeholder
-        if (state.messageHtml.isEmpty()) {
-            Text(
-                placeholder,
-                style = defaultTypography.copy(
-                    color = ElementTheme.colors.textSecondary,
+            RichTextEditor(
+                state = state,
+                registerStateUpdates = !subcomposing,
```
Comment (jonnyandrew) [inline libraries/textcomposer/impl/src/main/kotlin/io/element/android/libraries/textcomposer/TextComposer.kt:301]: Disable most of the editor functionality if it's just being measured for a subcomposition. This prevents it gaining focus and mutating the state.

**Comment by du82** [comment]
It would be cool if there was a way to permanently enable the rich text editor instead of having to tap two things for it

File: libraries/textcomposer/impl/src/main/kotlin/io/element/android/libraries/textcomposer/TextComposer.kt (hunk 2 above)
Comment (bmarty) [reply libraries/textcomposer/impl/src/main/kotlin/io/element/android/libraries/textcomposer/TextComposer.kt:301]: OK, thanks. Maybe add this comment also in the code?

**Review by bmarty — APPROVED** [review]
LGTM, thanks!

**Pull request merged**